build.assets/tooling/cmd/update-plist-version/main.go

```diff
@@ -0,0 +1,140 @@
+// positive integers separated by dots. A semver with a pre-release tag is not
+// valid.
+//
+// This is intended to be used on the tsh.app Info.plist files. Standard
+// releases have the required version number. Pre-releases do not.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"howett.net/plist"
+)
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <version> <plist-file>...\n", filepath.Base(os.Args[0]))
+		os.Exit(1)
+	}
+
+	// A version can only be three positive integers separated by periods.
+	versionRE := regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+$`)
+	version := os.Args[1]
+	if !versionRE.MatchString(version) {
+		version = "1.0"
+	}
+
+	for _, filename := range os.Args[2:] {
+		err := replaceFile(filename, func(in, out *os.File) error {
+			return updateVersion(version, in, out)
+		})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "could not update version of %s: %v\n", filename, err)
+			os.Exit(1)
+		}
+	}
+}
+
+func updateVersion(version string, in, out *os.File) error {
+	dict := map[string]any{}
+	d := plist.NewDecoder(in)
+	if err := d.Decode(&dict); err != nil {
+		return err
+	}
+
+	if _, ok := dict["CFBundleVersion"]; !ok {
+		return errors.New("CFBundleVersion not in plist file")
+	}
+	if _, ok := dict["CFBundleShortVersionString"]; !ok {
+		return errors.New("CFBundleShortVersionString not in plist file")
+	}
+	dict["CFBundleVersion"] = version
+	dict["CFBundleShortVersionString"] = version
+
+	e := plist.NewEncoder(out)
+	e.Indent("\t")
+	if err := e.Encode(dict); err != nil {
+		return err
+	}
+	// the plist encoder does not write a newline at the end of the last line.
+	_, err := out.Write([]byte{'\n'})
+	return err
+}
+
+// Replace calls fn to replace the contents of filename. fn is passed
+// the input file and the output file and should return when it has
+// finished writing to out or an error occurs. The file will be
+// replaced atomically on success, or will be left untouched on
+// error. File permission of filename is retained, ownership and
+// attributes are not.
+func replaceFile(filename string, fn func(in, out *os.File) error) (rerr error) {
+	inf, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	// defer inf.Close() and only return the error if not returning another error.
+	defer func() {
+		if err := inf.Close(); err != nil && rerr == nil {
+			rerr = err
+		}
+	}()
+
+	outdir := filepath.Dir(filename)
+	outname := filepath.Base(filename)
+	outf, err := os.CreateTemp(outdir, outname)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// If we're returning an error we need to delete the temp file
+		if rerr != nil {
+			// Print any os.Remove() error on stderr as we are already returning an error
+			if err := os.Remove(outf.Name()); err != nil {
+				fmt.Fprintf(os.Stderr, "could not remove temp file %s: %v\n", outf.Name(), err)
+			}
+		}
+		if err := outf.Close(); err != nil && rerr == nil {
```

Comment:
Close before the Remove? If a flush happens, I'm afraid it'll just bring the file back to life.

Reply:
I don't understand this. You can remove an open file, and it will not come back with any sort of flush. Also that will complicate the return error handling even more.